libtiledbsoma/test/unit_soma_group.cc

```diff
@@ -0,0 +1,226 @@
+    const std::string& uri_in,
+    Context& ctx,
+    int num_cells_per_fragment = 10,
+    int num_fragments = 1,
+    bool overlap = false,
+    bool allow_duplicates = false,
+    uint64_t timestamp = 1,
+    bool reuse_existing = false) {
+    std::string uri = fmt::format(
+        "{}-{}-{}-{}-{}",
+        uri_in,
+        num_cells_per_fragment,
+        num_fragments,
+        overlap,
+        allow_duplicates);
+    // Create array, if not reusing the existing array
+    if (!reuse_existing) {
+        auto vfs = VFS(ctx);
+        if (vfs.is_dir(uri)) {
+            vfs.remove_dir(uri);
+        }
+
+        // Create schema
+        ArraySchema schema(ctx, TILEDB_SPARSE);
+
+        auto dim = Dimension::create<int64_t>(
+            ctx, "d0", {0, std::numeric_limits<int64_t>::max() - 1});
+
+        Domain domain(ctx);
+        domain.add_dimension(dim);
+        schema.set_domain(domain);
+
+        auto attr = Attribute::create<int>(ctx, "a0");
+        schema.add_attribute(attr);
+        schema.set_allows_dups(allow_duplicates);
+        schema.check();
+
+        // Create array
+        Array::create(uri, schema);
+    }
+
+    // Open array for writing
+    Array array(ctx, uri, TILEDB_WRITE, timestamp);
+    if (LOG_DEBUG_ENABLED()) {
+        array.schema().dump();
+    }
+
+    // Generate fragments in random order
+    std::vector<int> frags(num_fragments);
+    std::iota(frags.begin(), frags.end(), 0);
+    std::shuffle(frags.begin(), frags.end(), std::random_device{});
+
+    for (auto i : frags) {
+        std::vector<int64_t> d0(num_cells_per_fragment);
+        for (int j = 0; j < num_cells_per_fragment; j++) {
+            // Overlap odd fragments when generating overlaps
+            if (overlap && i & 1) {
+                d0[j] = j + num_cells_per_fragment * (i - 1);
+            } else {
+                d0[j] = j + num_cells_per_fragment * i;
+            }
+        }
+        std::vector<int> a0(num_cells_per_fragment, i);
+
+        // Write data to array
+        Query query(ctx, array);
+        query.set_layout(TILEDB_UNORDERED)
+            .set_data_buffer("d0", d0)
+            .set_data_buffer("a0", a0);
+        query.submit();
+    }
+
+    array.close();
+
+    uint64_t nnz = num_fragments * num_cells_per_fragment;
+
+    if (allow_duplicates) {
+        return {uri, nnz};
+    }
+
+    // Adjust nnz when overlap is enabled
+    if (overlap) {
+        nnz = (num_fragments + 1) / 2 * num_cells_per_fragment;
+    }
+
+    return {uri, nnz};
+}
+
+};  // namespace
+
+TEST_CASE("SOMAGroup: basic") {
+    auto ctx = std::make_shared<Context>();
+
+    std::string uri_main_group = "mem://main-group";
+    Group::create(*ctx, uri_main_group);
+
+    std::string uri_sub_group = "mem://sub-group";
+    Group::create(*ctx, uri_sub_group);
+
+    auto [uri_sub_array, expected_nnz] = create_array("mem://sub-array", *ctx);
+
+    auto sg = SOMAGroup::open(TILEDB_WRITE, ctx, uri_main_group, "metadata", 1);
```

Comment:
What does sg stand for?

Reply:
`SOMAGroup`. I will rename to something more clear.